Python/chapter02/2.4 - Partition/miguel_2.4_soln.py

```diff
@@ -137,23 +151,28 @@ def partition_ll(ll: LinkedList, pivot: int) -> LinkedList:
     elements less than x. They do not need to appear
     in between the left and right partitions.
     Runtime: O(n)
-    Space Complexity: O(n)
+    Space Complexity: O(1)
     :param ll: an input linked list
     :param pivot: a number to partition around
     :return: a linked list that is 'partitioned'
     """
-    left_partition = []  # will contain values < pivot
-    right_partition = []  # will contain values >= pivot
+    left_partition = LinkedList()  # will contain values < pivot
+    right_partition = LinkedList()  # will contain values >= pivot
 
     n = ll.head
     while n is not None:
         if n.data < pivot:
-            left_partition.append(n.data)
+            left_partition.append_to_tail(n)
         else:
-            right_partition.append(n.data)
+            right_partition.append_to_tail(n)
         n = n.next
+    # last element may still be pointing
+    # to earlier elements that are less than
+    # pivot so we need to cut that link
+    right_partition.append_to_tail(None)
```

Comment:
can you instead add a `def pop_head(self) -> Node` to avoid doing this? that way you can do

```python
left_partition.append_to_tail(ll.pop_head())
```

that way you can also avoid touching the `head` and `next` pointers.